retrieval/discovery/consul.go

```diff
@@ -191,7 +184,7 @@ func (cd *ConsulDiscovery) watchServices(update chan<- *consulService) {
 		})
 		if err != nil {
 			log.Errorf("Error refreshing service list: %s", err)
-			<-time.After(consulRetryInterval)
+			time.Sleep(consulRetryInterval)
```

Comment:
How about also returning here when `done` is closed via a `select`? That way it doesn't take 15 seconds to shut down the target provider in the unlucky case.